core/store/src/trie/mod.rs

```diff
@@ -1118,36 +1178,37 @@ impl Trie {
         mode: KeyLookupMode,
     ) -> Result<Option<ValueRef>, StorageError> {
         let use_flat_storage =
-            matches!(mode, KeyLookupMode::FlatStorage) && self.flat_storage_chunk_view.is_some();
-
+            mode == KeyLookupMode::FlatStorage && self.flat_storage_chunk_view.is_some();
+        let charge_gas_for_trie_node_access =
+            mode == KeyLookupMode::Trie || self.charge_gas_for_trie_node_access;
         if use_flat_storage {
-            let value_from_flat_storage = self
-                .flat_storage_chunk_view
-                .as_ref()
-                .unwrap()
-                .get_value(&key)?
-                .map(|value| value.to_value_ref());
-            if self.recorder.is_some() {
-                // If recording, we need to look up in the trie as well to record the trie nodes,
-                // as they are needed to prove the value. Also, it's important that this lookup
-                // is done even if the key was not found, because intermediate trie nodes may be
-                // needed to prove the non-existence of the key.
-                let key_nibbles = NibbleSlice::new(key);
-                let value_from_trie = self.lookup(key_nibbles, false)?;
-                assert_eq!(&value_from_flat_storage, &value_from_trie);
-            }
-            Ok(value_from_flat_storage)
+            Ok(self.lookup_from_flat_storage(key, true)?.map(|value| value.to_value_ref()))
+        } else {
+            self.lookup_from_state_column(NibbleSlice::new(key), charge_gas_for_trie_node_access)
+        }
```

Comment:
nit: isn't it simpler to check use_flat_storage and return right away instead of evaluating charge_gas_for_trie_node_access?

Reply:
Ah, the boolean is used in another branch once I add the in-memory trie changes here. That branch goes before flat storage.